Sources/Web3Core/KeystoreManager/EtherscanTransactionChecker.swift

```diff
@@ -0,0 +1,49 @@
+//
+//  EtherscanTransactionChecker.swift
+//  Created by albertopeam on 28/12/22.
+//
+
+import Foundation
+import _Concurrency
+
+public struct EtherscanTransactionChecker: TransactionChecker {
+    private let urlSession: URLSessionProxy
+    private let apiKey: String
+
+    public init(urlSession: URLSessionProxy, apiKey: String) {
+        self.urlSession = urlSession
+        self.apiKey = apiKey
+    }
+
+    public func hasTransactions(address: String) async throws -> Bool {
+        let urlString = "https://api.etherscan.io/api?module=account&action=txlist&address=\(address)&startblock=0&page=1&offset=1&sort=asc&apikey=\(apiKey)"
+        guard let url = URL(string: urlString) else {
+            throw EtherscanTransactionCheckerError.invalidUrl
+        }
+        let request = URLRequest(url: url)
+        let result = try await urlSession.data(request: request)
+        let response = try JSONDecoder().decode(Response.self, from: result.0)
+        return !response.result.isEmpty
+    }
+}
+
+extension EtherscanTransactionChecker {
+    struct Response: Codable {
+        let result: [Transaction]
+    }
+    struct Transaction: Codable {}
+}
+
+public enum EtherscanTransactionCheckerError: Error {
+    case invalidUrl
+}
+
+public protocol URLSessionProxy {
+    func data(request: URLRequest) async throws -> (Data, HTTPURLResponse)
+}
+
+extension URLSession: URLSessionProxy {
+    public func data(request: URLRequest) async throws -> (Data, HTTPURLResponse) {
+        return try await data(for: request)
```

Comment:
The `data(for: ...)` is a custom extension function we have in the library.
There is a similar function in the API of the `URLSession` itself but it just returns `URLResponse` instead of the `HTTPURLResponse` as the second value of the tuple.

What I suggest is that you could try to replace protocol and this extension with the following:
```
public protocol URLSessionProxy {
    func data(for: URLRequest) async throws -> (Data, URLResponse)
}

extension URLSession: URLSessionProxy { }
```

We will eventually remove `data(for: ...)` that returns `(Data, HTTPURLResponse)`.

Reply:
I will review the existence of this function

Reply:
I have took a look but I got blocked as I can not use test doubles because the Apple and Web3 APIs use function extensions, so I can't override them in the Mock.

This code compiles
```swift
public protocol URLSessionProxy: URLSession {
    func data(for request: URLRequest) async throws -> (Data, URLResponse)
}

extension URLSession: URLSessionProxy {}
```

This one not, error msg: "Non-@objc instance method 'data(for:)' is declared in extension of 'URLSession' and cannot be overridden"
```swift
final class URLSessionMock: URLSession {
    var response: (Data, HTTPURLResponse) = (Data(), HTTPURLResponse())
    
    func data(for request: URLRequest) async throws -> (Data, URLResponse) {
        return response
    }
}
```

Reply:
If you know some trick to avoid using the protocol I will study it :)

Reply:
Last minute idea, expose URLsession publicly and then wrap with our Proxy, having an internal init using directly the proxy to be able to test easily the EtherscanTransactionChecker.
Best to check the code, what do you think?
```swift
internal protocol URLSessionProxy {
    func data(for request: URLRequest) async throws -> (Data, URLResponse)
}
internal struct URLSessionProxyImplementation: URLSessionProxy {
    let urlSession: URLSession
    
    func data(for request: URLRequest) async throws -> (Data, URLResponse) {
        try await urlSession.data(for: request)
    }
}

public struct EtherscanTransactionChecker: TransactionChecker {
    private let urlSession: URLSessionProxy
    private let apiKey: String

    public init(urlSession: URLSession, apiKey: String) {
        self.urlSession = URLSessionProxyImplementation(urlSession: urlSession)
        self.apiKey = apiKey
    }
    
    internal init(urlSession: URLSessionProxy, apiKey: String) {
        self.urlSession = urlSession
        self.apiKey = apiKey
    }
}
```

Reply:
This code compiles fine and should be mockable:
```swift
public protocol URLSessionProxy {
    func data(for request: URLRequest) async throws -> (Data, URLResponse)
}

extension URLSession: URLSessionProxy {}

public struct EtherscanTransactionChecker {
    private let urlSession: URLSessionProxy
    private let apiKey: String

    public init(urlSession: URLSessionProxy = URLSession(configuration: .default), apiKey: String) {
        self.urlSession = urlSession
        self.apiKey = apiKey
    }
}
```

You do not need to override anything in `URLSessionProxyImplementation` for tests. You do not need the `URLSessionProxyImplementation` even. Instead, you should keep using your mock but change the `HTTPURLResponse` to `URLResponse`:
```swift
final class URLSessionMock: URLSessionProxy {
    var response: (Data, HTTPURLResponse) = (Data(), HTTPURLResponse())

    func data(request: URLRequest) async throws -> (Data, HTTPURLResponse) {
        return response
    }
}
```

Your attempt to use `URLSessionProxy` in this way was wrong conceptually:
```swift
public protocol URLSessionProxy: URLSession {
    func data(for request: URLRequest) async throws -> (Data, URLResponse)
}

extension URLSession: URLSessionProxy {}
```

But in this statement `protocol URLSessionProxy: URLSession` you were saying that only classes that conform to `URLSession` can implement this protocol, which won't work.

Try the code I shared above. Thanks for your effort!

Reply:
I appreciate your words!, thanks. Happy to be coding here :)

The code you shared is not compiling on my mac(using xcode13.4.1). Maybe I understood it wrong :)
If you know why is not compiling please let me know to study your approach :)
<img width="1468" alt="Captura de Pantalla 2023-01-15 a las 22 35 24" src="https://user-images.githubusercontent.com/912381/212568612-e117ccc2-f2b2-4d6c-9e80-e91b6de5080b.png">

My previous idea was to make the proxy as internal to avoid exposing the type because we want to test EtherscanTransactionChecker, so I think is a small piece of code but much cleaner for the devs.
I didn't want to override `URLSessionProxyImplementation` in tests, I used it to provide a unified API between URLSession and URLSessionProxy, both have their own init in EtherscanTransactionChecker and only URLSession is exposed to devs. what do you think?
```swift
public struct EtherscanTransactionChecker: TransactionChecker {
    private let urlSession: URLSessionProxy
    private let apiKey: String

    public init(urlSession: URLSession, apiKey: String) {
        self.urlSession = URLSessionProxyImplementation(urlSession: urlSession)
        self.apiKey = apiKey
    }
    
    internal init(urlSession: URLSessionProxy, apiKey: String) {
        self.urlSession = urlSession
        self.apiKey = apiKey
    }
}
```
PROS vs CONS of the strategy:
PROS:
* not exposing a protocol that mimics URLSession and is an internal detail for the tests
CONS:
* A bit more of code to hide it from devs using EtherscanTransactionChecker

Anyway, If I made any error on the snippet that you have shared I will study your solution, but I think that the one provided here has some value also :)